### PR TITLE
Convert to number

### LINKED ISSUE #60: convert_to_number
very simple.. you have a string, just make it a number

input"23"
output 23

### PR DESCRIPTION
**Pull request by ArwaAbdelkhalik**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [ ] The function has correct type annotations
- [ ] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in any docstrings or tests
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] There are no commented lines of code
- [ ] There are no `print` statements anywhere
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/convert_to_number.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+""" "
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:1]: 1- There is  more " 
2- The file name was not describes the function's behavior clearly
3- The a module docstring in the function file did not complete

File: solutions/tests/test_convert_to_number.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+""" "
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:1]: 1- There is more "
2- The a module docstring in the function file did not complete

File: solutions/tests/test_convert_to_number.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+        >>> convert_to_number("3.14")
+        3.14
+
+    Raises:
+        ValueError: If the input string cannot be converted to a number.
+    """
+    if not isinstance(input_str, str):
+        raise ValueError("Input must be a string.")
+
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:32]: all this commands until before class for function files ,not in test files

File: solutions/tests/test_convert_to_number.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            return float(input_str)
+        return int(input_str)
+    except ValueError:
+        raise ValueError(f"Cannot convert '{input_str}' to a number.")
+
+
+class TestConvertToNumber(unittest.TestCase):
+    """Unit tests for the convert_to_number function.
+
+    The tests verify:
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:44]: all test verify must be in module docstrings place not here

File: solutions/tests/test_convert_to_number.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    - Conversion of valid integer strings to integers.
+    - Conversion of valid float strings to floats.
+    - Handling of invalid strings (non-numeric).
+    - Error handling for non-string inputs.
+    """
+
+    def test_valid_integer_string(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:51]: 1-Every unit test has only one assertion

2- test method names are too generic

File: solutions/tests/test_convert_to_number.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        """Test conversion of valid integer strings."""
+        self.assertEqual(convert_to_number("42"), 42)
+        self.assertEqual(convert_to_number("-7"), -7)
+
+    def test_valid_float_string(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:56]: 1-Every unit test has only one assertion

2- test method names are too generic

File: solutions/tests/test_convert_to_number.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        """Test conversion of valid float strings."""
+        self.assertEqual(convert_to_number("3.14"), 3.14)
+        self.assertEqual(convert_to_number("-0.001"), -0.001)
+
+    def test_invalid_string(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:61]: Every unit test has only one assertion

File: solutions/tests/test_convert_to_number.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        """Test behavior with invalid strings."""
+        with self.assertRaises(ValueError):
+            convert_to_number("abc")
+        with self.assertRaises(ValueError):
+            convert_to_number("123abc")
+
+    def test_non_string_input(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:68]: 1-Every unit test has only one assertion

2- test method names are too generic

File: solutions/convert_to_number.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        float: If the string represents a float.
+
+    Examples:
+        >>> convert_to_number("42")
+        42
+
+        >>> convert_to_number("3.14")
+        3.14
+
+    Raises:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:25]: the place of the raises before examples

File: solutions/convert_to_number.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    """Converts a string to a number (integer or float).
+
+    Parameters:
+        input_str (str): The string to convert to a number.
+
+    Returns:
+        int: If the string represents an integer.
+        float: If the string represents a float.
+
+    Examples:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:18]: must be includes 3 or more (passing!) doctests

File: solutions/convert_to_number.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
+""" "
+Created on 0084/01/2025
+@author: Arwa Mohamed
+
+"""
+
+
+def convert_to_number(input_str):
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:8]: The function has not type annotations

File: solutions/convert_to_number.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        >>> convert_to_number("42")
+        42
+
+        >>> convert_to_number("3.14")
+        3.14
+
+    Raises:
+        ValueError: If the input string cannot be converted to a number.
+    """
+    if not isinstance(input_str, str):
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:28]: miss comments are clear and describe the strategy

File: solutions/convert_to_number.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        raise ValueError("Input must be a string.")
+
+    try:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:31]: miss comments are clear and describe the strategy

File: solutions/convert_to_number.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        if "." in input_str:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:32]: miss comments are clear and describe the strategy

File: solutions/convert_to_number.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+            return float(input_str)
+        return int(input_str)
+    except ValueError:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:35]: miss comments are clear and describe the strategy

File: solutions/tests/test_convert_to_number.py [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    The tests verify:
+    - Conversion of valid integer strings to integers.
+    - Conversion of valid float strings to floats.
+    - Handling of invalid strings (non-numeric).
+    - Error handling for non-string inputs.
+    """
+
+    def test_valid_integer_string(self):
+        """Test conversion of valid integer strings."""
+        self.assertEqual(convert_to_number("42"), 42)
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:53]: Missing explicit tests for boundary cases such as "24", "-7"

File: solutions/tests/test_convert_to_number.py [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+"""
+
+import unittest
+
+
+def convert_to_number(input_str):
+    """Converts a string to a number (integer or float).
+
+    Parameters:
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:13]: Does not clarify additional assumptions about valid inputs (e.g., numeric characters, expected format).

File: solutions/tests/test_convert_to_number.py [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+
+    def test_non_string_input(self):
+        """Test error handling for non-string inputs."""
+        with self.assertRaises(ValueError):
+            convert_to_number(123)
+        with self.assertRaises(ValueError):
+            convert_to_number(None)
+        with self.assertRaises(ValueError):
+            convert_to_number([1, 2, 3])
+
```
Comment (AzzaOmer1) [inline solutions/tests/test_convert_to_number.py:76]: No explicit tests for boundary cases like empty strings or strings with leading/trailing spaces.

File: solutions/convert_to_number.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+Created on 0084/01/2025
+@author: Arwa Mohamed
+
+"""
+
+
+def convert_to_number(input_str):
+    """Converts a string to a number (integer or float).
+
+    Parameters:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:11]: Other assumptions: Does not specify assumptions like whether empty strings or strings with spaces are valid.

File: solutions/convert_to_number.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+        input_str (str): The string to convert to a number.
+
+    Returns:
```
Comment (AzzaOmer1) [inline solutions/convert_to_number.py:14]: Other assumptions: Does not clarify that it raises exceptions for invalid input.

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]
Edit all issues
